src/toxcore/dht/server/mod.rs

```diff
@@ -208,9 +216,21 @@ impl Server {
             lan_discovery_enabled: true,
             is_ipv6_enabled: false,
             initial_bootstrap: Vec::new(),
+            precomputed_keys: Arc::new(RwLock::new(LruCache::new(PRECOMPUTED_LRU_CACHE_SIZE))),
         }
     }
 
+    /// Get PrecomputedKey of a peer or newly make and save it.
+    fn get_precomputed_key(&self, pk: PublicKey) -> PrecomputedKey {
+        let mut precomputed_cache = self.precomputed_keys.write();
```

Comment:
Better optimize it as:

1. try read lock first
2. if cache for PK has a value -> return it

3. if cache is empty
4. close read lock
5. open write lock
....

Reply:
https://docs.rs/lru/0.1.9/lru/struct.LruCache.html#method.get

Even `get()` requires mutable.

Reply:
But https://docs.rs/lru/0.1.9/lru/struct.LruCache.html#method.contains does not.

@kurnevsky is this optimization worth implementing?

Reply:
contains returns `bool`.
if `contains` returns `true` then open write lock.
if `contains` returns `false` then open write lock.

Reply:
if contains returns false you can compute PrecomputedKey and after that take write lock. That will reduce the time you spend in the critical section. But we'd better benchmark it.

Reply:
@kpp this cache takes `&mut self` on get because it tracks frequency for each element how often you get it so it can evict the less used elements. So it means that we have to acquire write lock if we want to either get or add a new key. And I think one lock is preferable because acquiring a lock should be slow operation. Minimizing instructions into critical section shouldn't worth it.

Reply:
Ok